src/signTx.c

```diff
@@ -1665,6 +1779,194 @@ static void signTx_handleScriptDataHashAPDU(uint8_t p2, uint8_t* wireDataBuffer,
+static void signTx_handleCollateral_ui_runStep()
+{
+	TRACE("UI step %d", ctx->ui_step);
+	ui_callback_fn_t* this_fn = signTx_handleCollateral_ui_runStep;
+
+	UI_STEP_BEGIN(ctx->ui_step, this_fn);
+
+	UI_STEP(HANDLE_COLLATERAL_STEP_DISPLAY) {
+		char headerText[18];
+		snprintf(headerText, SIZEOF(headerText), "Collateral #%u", BODY_CTX->currentCollateral);
+		ui_displayInputScreen(headerText, &BODY_CTX->stageData.collateral, this_fn);
+	}
+
+	UI_STEP(HANDLE_COLLATERAL_STEP_RESPOND) {
+		respondSuccessEmptyMsg();
+
+		// Advance stage to the next input
+		ASSERT(BODY_CTX->currentCollateral < ctx->numCollaterals);
+		BODY_CTX->currentCollateral++;
+
+		if (BODY_CTX->currentCollateral == ctx->numCollaterals) {
+			advanceStage();
+		}
+	}
+	UI_STEP_END(HANDLE_COLLATERAL_STEP_INVALID);
+}
+
+__noinline_due_to_stack__
+static void signTx_handleCollateralAPDU(uint8_t p2, uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	TRACE_STACK_USAGE();
+	{
+		// sanity checks
+		CHECK_STAGE(SIGN_STAGE_BODY_COLLATERALS);
+		ASSERT(BODY_CTX->currentCollateral < ctx->numCollaterals);
+
+		VALIDATE(p2 == P2_UNUSED, ERR_INVALID_REQUEST_PARAMETERS);
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+
+	BODY_CTX->stageData.collateral = extractTransactionInput(wireDataBuffer, wireDataSize);
+
+	security_policy_t policy = policyForSignTxCollaterals(ctx->commonTxData.txSigningMode);
+	TRACE("Policy: %d", (int) policy);
+	ENSURE_NOT_DENIED(policy);
+
+	{
+		// add to tx
+		TRACE("Adding collateral to tx hash");
+		txHashBuilder_addCollateral(
+		        &BODY_CTX->txHashBuilder,
+		        BODY_CTX->stageData.collateral.txHashBuffer, SIZEOF(BODY_CTX->stageData.collateral.txHashBuffer),
+		        BODY_CTX->stageData.collateral.parsedIndex
+		);
+	}
+
+	{
+		// select UI steps
+		switch (policy) {
+#	define  CASE(POLICY, UI_STEP) case POLICY: {ctx->ui_step=UI_STEP; break;}
+			CASE(POLICY_SHOW_BEFORE_RESPONSE, HANDLE_COLLATERAL_STEP_DISPLAY);
+			CASE(POLICY_ALLOW_WITHOUT_PROMPT, HANDLE_COLLATERAL_STEP_RESPOND);
+#	undef   CASE
+		default:
+			THROW(ERR_NOT_IMPLEMENTED);
+		}
+	}
+	signTx_handleCollateral_ui_runStep();
+}
+
+// ========================= REQUIRED SIGNERS ===========================
+
+enum {
+	HANDLE_REQUIRED_SIGNERS_STEP_DISPLAY = 1400,
+	HANDLE_REQUIRED_SIGNERS_STEP_RESPOND,
+	HANDLE_REQUIRED_SIGNERS_STEP_INVALID,
+};
+
+static void signTx_handleRequiredSigner_ui_runStep()
+{
+	TRACE("UI step %d", ctx->ui_step);
+	ui_callback_fn_t* this_fn = signTx_handleRequiredSigner_ui_runStep;
+
+	UI_STEP_BEGIN(ctx->ui_step, this_fn);
+
+	UI_STEP(HANDLE_REQUIRED_SIGNERS_STEP_DISPLAY) {
+		switch (BODY_CTX->stageData.requiredSigner.type) {
+		case REQUIRED_SIGNER_WITH_PATH:
+			ui_displayPathScreen("Required signer", &BODY_CTX->stageData.requiredSigner.keyPath, this_fn);
+			break;
+		case REQUIRED_SIGNER_WITH_HASH:
+			ui_displayHexBufferScreen("Required signer", BODY_CTX->stageData.requiredSigner.keyHash, SIZEOF(BODY_CTX->stageData.requiredSigner.keyHash), this_fn);
```

Comment:
I'm not sure whether to use hex, so I asked IOHK: https://input-output-rnd.slack.com/archives/G010N9UFDNH/p1637576833039200